### PR TITLE
Add unbuffer support for impala

### PR DESCRIPTION
**Pull request by kuszz**
Address #13732

### What changes are proposed in this pull request?

Add canUnBuffer ability on stream which BaseFileSystem use

### Why are the changes needed?

The impala need canunbuffer ability to cache file handles

### Does this PR introduce any user facing changes?

User can unbuffer the some streams

### CODE REVIEW

File: core/client/fs/src/main/java/alluxio/client/block/stream/BlockInStream.java [hunk 1]
```diff
@@ -32,6 +32,7 @@
 
 import com.google.common.annotations.VisibleForTesting;
 import com.google.common.base.Preconditions;
+import org.apache.hadoop.fs.CanUnbuffer;
```
Comment (apc999) [inline core/client/fs/src/main/java/alluxio/client/block/stream/BlockInStream.java:35]: in `core/client/fs` module, we shall not introduce a dependency to hadoop which will cause dependencies issues.
Module `core/client/hdfs` is the wrapper on top of Alluxio In/out streams and make them HDFS compatible.

File: core/client/fs/src/main/java/alluxio/client/block/stream/BlockInStream.java [hunk 2]
```diff
@@ -520,4 +521,13 @@ public BlockInStreamSource getSource() {
   public long getId() {
     return mId;
   }
+
+  @Override
```
Comment (apc999) [inline core/client/fs/src/main/java/alluxio/client/block/stream/BlockInStream.java:525]: Just claim it without introducing a interface
```suggestion
```

File: core/client/fs/src/main/java/alluxio/client/block/stream/BlockInStream.java [hunk 3]
```diff
@@ -520,4 +521,13 @@ public BlockInStreamSource getSource() {
...
+  public void unbuffer() {
+    try {
+      closeDataReader();
+    } catch (IOException e) {
+      LOG.error("Failed to close DataReader: {}", e.getMessage());
```
Comment (apc999) [inline core/client/fs/src/main/java/alluxio/client/block/stream/BlockInStream.java:530]:
```suggestion
      LOG.error("Failed to close DataReader", e);
```

File: core/client/fs/src/main/java/alluxio/client/file/AlluxioFileInStream.java [hunk 4]
```diff
@@ -33,6 +33,9 @@
 import com.google.common.base.Preconditions;
 import com.google.common.base.Supplier;
 import com.google.common.io.Closer;
+import org.apache.hadoop.fs.CanUnbuffer;
+import org.apache.hadoop.fs.StreamCapabilities;
+import org.apache.hadoop.util.StringUtils;
```
Comment (apc999) [inline core/client/fs/src/main/java/alluxio/client/file/AlluxioFileInStream.java:38]: same here, no hadoop dependency here.

File: core/client/hdfs/src/main/java/alluxio/hadoop/HdfsFileInputStream.java [hunk 5]
```diff
@@ -39,7 +43,7 @@
  */
 @NotThreadSafe
 public class HdfsFileInputStream extends InputStream implements Seekable, PositionedReadable,
-    ByteBufferReadable {
+    ByteBufferReadable, StreamCapabilities, CanUnbuffer {
```
Comment (apc999) [inline core/client/hdfs/src/main/java/alluxio/hadoop/HdfsFileInputStream.java:46]: do we know staring from which hadoop version , these interfaces are introduced?

File: core/client/fs/pom.xml [hunk 6]
```diff
@@ -92,6 +92,12 @@
       <type>test-jar</type>
       <scope>test</scope>
     </dependency>
+    <dependency>
+      <groupId>org.apache.hadoop</groupId>
+      <artifactId>hadoop-common</artifactId>
+      <version>3.2.1</version>
+      <scope>compile</scope>
+    </dependency>
```
Comment (apc999) [inline core/client/fs/pom.xml:100]: No hadoop dependency shall be added to `core/client/fs/pom.xml`
```suggestion
```

**Review by apc999 — CHANGES_REQUESTED** [review]
Comments and questions are left

File: core/client/hdfs/src/main/java/alluxio/hadoop/HdfsFileInputStream.java (hunk 5 above)
Comment (maobaolong) [reply core/client/hdfs/src/main/java/alluxio/hadoop/HdfsFileInputStream.java:46]: https://issues.apache.org/jira/browse/HDFS-7694 
This ticket show that it string from 2.6.4

File: core/client/fs/pom.xml (hunk 6 above)
Comment (maobaolong) [reply core/client/fs/pom.xml:100]: +1

**Review by maobaolong — CHANGES_REQUESTED** [review]
@kuszz Thanks for this feature, but please keep no hadoop dependency is it was. So think about to move the canbuffer logic to the underfs/hdfs module or find some other approach instead of involve hadoop dependency to core module.

File: core/client/fs/src/main/java/alluxio/client/block/stream/BlockInStream.java (hunk 1 above)
Comment (kuszz) [reply core/client/fs/src/main/java/alluxio/client/block/stream/BlockInStream.java]: Got it, I have modified code for these issues

File: core/client/hdfs/src/main/java/alluxio/hadoop/HdfsFileInputStream.java (hunk 5 above)
Comment (apc999) [reply core/client/hdfs/src/main/java/alluxio/hadoop/HdfsFileInputStream.java:45]: then one problem will be: if we build Alluxio src code with Hadoop version < 2.6.4, the build will fail.
is it possible that we add `hasCapability` and `unbuffer` here, but without claiming implementing
interfaces `StreamCapabilities`, `CanUnbuffer`? will that work for Impala? e.g., does Impala checks if the InputStream extends these interfaces explicitly?

**Review by apc999 — CHANGES_REQUESTED** [review]
one question left about building this with Hadoop version < 2.6

File: core/client/hdfs/src/main/java/alluxio/hadoop/HdfsFileInputStream.java (hunk 5 above)
Comment (kuszz) [reply core/client/hdfs/src/main/java/alluxio/hadoop/HdfsFileInputStream.java:45]: @apc999 Ok, I will ask impala contributor who ask for this feature if can not check it

**Review by apc999 — CHANGES_REQUESTED** [review]
Can you add unit tests on the new method?

**Comment by apc999** [comment]
ping @kuszz

**Comment by kuszz** [comment]
> ping @kuszz

Sorry, I missed the github email，I will add UT as soon as possible

File: core/client/hdfs/src/main/java/alluxio/hadoop/HdfsFileInputStream.java (hunk 5 above)
Comment (apc999) [reply core/client/hdfs/src/main/java/alluxio/hadoop/HdfsFileInputStream.java:45]: as discussed offline, we can follow the same convention as introducing `ByteBufferReadable`

File: core/server/master/src/main/java/alluxio/master/metrics/MetricsStore.java [hunk 7]
```diff
@@ -224,7 +224,8 @@ public void initMetricKeys() {
             long total =
                 MetricsSystem.counter(MetricKey.CLUSTER_BYTES_READ_DIRECT.getName()).getCount()
                 + MetricsSystem.counter(MetricKey.CLUSTER_BYTES_READ_REMOTE.getName()).getCount()
-                + MetricsSystem.counter(MetricKey.CLUSTER_BYTES_READ_DOMAIN.getName()).getCount();
+                + MetricsSystem.counter(MetricKey.CLUSTER_BYTES_READ_DOMAIN.getName()).getCount()
+                + MetricsSystem.counter(MetricKey.CLUSTER_BYTES_READ_LOCAL.getName()).getCount();
```
Comment (apc999) [inline core/server/master/src/main/java/alluxio/master/metrics/MetricsStore.java:228]: this change is unrelated and already merged in by da3210a470025172dc9d49ea246909ca61305549, right?

Comment (kuszz) [reply core/server/master/src/main/java/alluxio/master/metrics/MetricsStore.java:228]: Yes, sorry, I merged local master and add this change by accident, should I do something for this?

**Review by apc999 — APPROVED** [review]

**Comment by apc999** [comment]
@ggezer  any input on this PR?

File: core/client/hdfs/src/main/java/alluxio/hadoop/HdfsFileInputStream.java (hunk 5 above)
Comment (apc999) [reply core/client/hdfs/src/main/java/alluxio/hadoop/HdfsFileInputStream.java:45]: @ggezer  in short, this PR is not introducing any new change in backward compatibility when building Alluxio.

This is because `org.apache.hadoop.fs.ByteBufferReadable` (introduced a few years back into `HdfsFileInputStream`) already breaks the build. As a result, if someone wants to build the Alluxio with north-bound HDFS client to be 2.2 or 1.x, they will already fail. Actually the two interfaces `StreamCapabilities`, `CanUnbuffer` is introduced to Hadoop earlier than `ByteBufferReadable`, so this PR is not making this situation  worse. 

Thus I think this change shall be fine, given we don't even notice this break in the last few years --- largely meaning no one attempts to build Alluxio north bound to be HDFS 1.x or 2.2. So we shall be good.

**Comment by apc999** [comment]
@ggezer I addressed your qestion https://github.com/Alluxio/alluxio/pull/13733/files#r680319496. PTAL

**Review by ggezer — APPROVED** [review]

**Comment by apc999** [comment]
alluxio-bot, merge this please
